common/tor.c

```diff
@@ -58,175 +54,198 @@ struct tor_service_reaching {
+		reach->noauth = true;
+	else if (strstr((char *)reach->buffer, "HASHEDPASSWORD")
+		 && (reach->ld->tor_service_password != NULL)) {
+		reach->noauth = false;
+		sprintf((char *)reach->cookie, "\"%s\"",
+			reach->ld->tor_service_password);
+	} else if ((p = strstr((char *)reach->buffer, "COOKIEFILE="))) {
+		assert(strlen(p) > 12);
+		p2 = strstr((char *)(p + 12), "\"");
+		assert(p2 != NULL);
+		i = strlen(p2);
+		*(char *)(p + (strlen(p) - i)) = 0;
+		int fd = open((char *)(p + 12), O_RDONLY);
+
+		if (fd < 0)
+			return io_close(conn);
+		if (!read(fd, &buf, sizeof(buf))) {
+			close(fd);
+			return io_close(conn);
+		} else
+			close(fd);
+
+		hex_encode(buf, 32, (char *)reach->cookie, 80);
+		reach->noauth = false;
+	} else
+		return io_close(conn);
 
-	return io_tor_connect_authenticate(conn,reach);
+	return io_tor_connect_authenticate(conn, reach);
 
 }
 
-static struct io_plan *io_tor_connect_after_protocolinfo(struct io_conn *conn, struct tor_service_reaching *reach)
+static struct io_plan *io_tor_connect_after_protocolinfo(struct io_conn *conn,
+							 struct
+							 tor_service_reaching
+							 *reach)
 {
 
-	memset(reach->buffer,0,MAX_TOR_SERVICE_READBUFFER_LEN);
-	return io_read_partial(conn, reach->buffer, MAX_TOR_SERVICE_READBUFFER_LEN ,&reach->hlen, &io_tor_connect_after_answer_pi, reach);
+	memset(reach->buffer, 0, MAX_TOR_SERVICE_READBUFFER_LEN);
+	return io_read_partial(conn, reach->buffer,
+			       MAX_TOR_SERVICE_READBUFFER_LEN - 1, &reach->hlen,
+			       &io_tor_connect_after_answer_pi, reach);
 }
 
-static struct io_plan *io_tor_connect_after_resp_to_connect(struct io_conn *conn, struct tor_service_reaching *reach)
+static struct io_plan *io_tor_connect_after_resp_to_connect(struct io_conn
+							    *conn,
+							    struct
+							    tor_service_reaching
+							    *reach)
 {
 
-			 sprintf((char *)reach->buffer, "PROTOCOLINFO 1\r\n");
-		  	 reach->hlen = strlen((char *)reach->buffer);
+	sprintf((char *)reach->buffer, "PROTOCOLINFO 1\r\n");
+	reach->hlen = strlen((char *)reach->buffer);
 
-	return io_write(conn, reach->buffer,  reach->hlen, io_tor_connect_after_protocolinfo, reach);
+	return io_write(conn, reach->buffer, reach->hlen,
+			io_tor_connect_after_protocolinfo, reach);
 }
 
-
 static struct io_plan *tor_connect_finish(struct io_conn *conn,
-		struct tor_service_reaching *reach)
+					  struct tor_service_reaching *reach)
 {
-	return io_tor_connect_after_resp_to_connect(conn,reach);
+	return io_tor_connect_after_resp_to_connect(conn, reach);
 };
 
-static struct io_plan *tor_conn_init(struct io_conn *conn, struct lightningd *ld)
+static struct io_plan *tor_conn_init(struct io_conn *conn,
+				     struct lightningd *ld)
 {
 	static struct tor_service_reaching reach;
 	static struct addrinfo *ai_tor;
 	reach.ld = ld;
-	getaddrinfo(tal_strdup(NULL, "127.0.0.1"),tal_strdup(NULL, "9051"), NULL,&ai_tor);
+	getaddrinfo(tal_strdup(NULL, "127.0.0.1"), tal_strdup(NULL, "9051"),
```

Comment:
Why do a `tal` with a `NULL` when the inputs are constants?  You are allocating memory unnecessarily and further more with a `NULL` parent they will not be freed, leaking memory.

Reply:
oooppsss ! they still are constants :-)